packages/tldraw/src/lib/ui/components/Toolbar/OverflowingToolbar.tsx

```diff
@@ -0,0 +1,199 @@
+import { areShortcutsDisabled } from '../../hooks/useKeyboardShortcuts'
+import { TLUiToolItem } from '../../hooks/useTools'
+import { useTranslation } from '../../hooks/useTranslation/useTranslation'
+import { TldrawUiButton } from '../primitives/Button/TldrawUiButton'
+import { TldrawUiButtonIcon } from '../primitives/Button/TldrawUiButtonIcon'
+import {
+	TldrawUiDropdownMenuContent,
+	TldrawUiDropdownMenuRoot,
+	TldrawUiDropdownMenuTrigger,
+} from '../primitives/TldrawUiDropdownMenu'
+import { TldrawUiMenuContextProvider } from '../primitives/menus/TldrawUiMenuContext'
+
+export const IsInOverflowContext = createContext(false)
+
+export function OverflowingToolbar({ children }: { children: React.ReactNode }) {
+	const editor = useEditor()
+	const id = useId().replace(/:/g, '_')
+	const breakpoint = useBreakpoint()
+	const msg = useTranslation()
+
+	const overflowIndex = Math.min(8, 5 + breakpoint)
+
+	const [totalItems, setTotalItems] = useState(0)
+	const mainToolsRef = useRef<HTMLDivElement>(null)
+	const [lastActiveOverflowItem, setLastActiveOverflowItem] = useState<string | null>(null)
+
+	const css = useMemo(() => {
+		const showInMainSelectors = []
+		const hideFromOverflowSelectors = []
+
+		if (lastActiveOverflowItem) {
+			showInMainSelectors.push(`[data-value="${lastActiveOverflowItem}"]`)
+		} else {
+			showInMainSelectors.push(`:nth-child(${overflowIndex + 1})`)
+		}
+
+		for (let i = 0; i < overflowIndex; i++) {
+			showInMainSelectors.push(`:nth-child(${i + 1})`)
+			hideFromOverflowSelectors.push(`:nth-child(${i + 1})`)
+		}
+
+		return `
+			#${id}_main > *:not(${showInMainSelectors.join(', ')}) {
+				display: none;
+			}
+			${hideFromOverflowSelectors.map((s) => `#${id}_more > *${s}`).join(', ')} {
+				display: none;
+			}
+        `
+	}, [lastActiveOverflowItem, id, overflowIndex])
+
+	const onDomUpdate = useEvent(() => {
+		if (!mainToolsRef.current) return
+
+		const children = Array.from(mainToolsRef.current.children)
+		setTotalItems(children.length)
+
+		// If the last active overflow item is no longer in the overflow, clear it
+		const lastActiveElementIdx = children.findIndex(
+			(el) => el.getAttribute('data-value') === lastActiveOverflowItem
+		)
+		if (lastActiveElementIdx <= overflowIndex) {
+			setLastActiveOverflowItem(null)
+		}
+
+		// But if there's a new active item...
+		const activeElementIdx = Array.from(mainToolsRef.current.children).findIndex(
+			(el) => el.getAttribute('aria-checked') === 'true'
+		)
+		if (activeElementIdx === -1) return
+
+		// ...and it's in the overflow, set it as the last active overflow item
+		if (activeElementIdx >= overflowIndex) {
+			setLastActiveOverflowItem(children[activeElementIdx].getAttribute('data-value'))
+		}
+	})
+
+	useLayoutEffect(() => {
+		onDomUpdate()
+	})
+
+	useLayoutEffect(() => {
+		if (!mainToolsRef.current) return
+
+		const mutationObserver = new MutationObserver(onDomUpdate)
+		mutationObserver.observe(mainToolsRef.current, {
```

Comment:
oo fancy!

Reply:
i'm obligated to put this here:
![mutation_observer](https://github.com/tldraw/tldraw/assets/469604/485961d5-6fba-4ed0-a830-60e035889f6d)